demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,448 @@
+         * character in the topic filter string. */
+        if( pTopicName[ nameIndex ] == pTopicFilter[ filterIndex ] )
+        {
+            /* Handle special corner cases regarding wildcards at the end of
+             * topic filters, as documented by the MQTT protocol spec. */
+            shouldStopMatching = matchEndWildcards( pTopicFilter,
+                                                    topicFilterLength,
+                                                    topicNameLength,
+                                                    nameIndex,
+                                                    filterIndex );
+            matchFound = shouldStopMatching;
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            shouldStopMatching = matchWildcards( pTopicFilter,
+                                                 pTopicName,
+                                                 topicNameLength,
+                                                 filterIndex,
+                                                 &nameIndex,
+                                                 &matchFound );
+        }
+
+        if( shouldStopMatching == true )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( matchFound == false )
+    {
+        /* If the end of both strings has been reached, they match. This represents the
+         * case when the topic filter contains a wildcard but neither at the starting nor
+         * the ending. For example, when matching "sport/+/player" topic filter with
+         * "sport/hockey/player" topic name. */
+        matchFound = ( ( nameIndex == topicNameLength ) &&
+                       ( filterIndex == topicFilterLength ) ) ? true : false;
+    }
+
+    return matchFound;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchTopic( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength )
+{
+    assert( pTopicName != NULL );
+    assert( topicNameLength != 0 );
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+
+    bool status = false;
+
+    /* Check for an exact match if the incoming topic name and the registered
+     * topic filter length match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 ) ? true : false;
+    }
+
+    /* If an exact match was not found, match against wildcard characters in topic filter.
+     * Note: According to the MQTT 3.1.1 specification, incoming PUBLISH topic names starting
+     * the "$" character cannot be matched against topic filter starting with a wildcard. */
+    if( ( status == false ) && ( ( pTopicName[ 0 ] != '$' ) &&
```

Comment:
Not sure I understand this condition. Are you trying to prevent the following:

```c
! ( ( pTopicName[ 0 ] == '$' ) &&  ( ( pTopicFilter[ 0 ] == '+' ) || ( pTopicFilter[ 0 ] == '#' ) ) )
```

i.e. If the topic on which the message is received, starts with `$` then it should match the registered topic filter starting with `#` or `+`.